### DISCUSSION TITLE
Why SQLAlchemy Disctinct is so slower than Raw Sql Distinct?

### DISCUSSION THREAD

**Discussion by gabrielmrts** [sqlalchemy/sqlalchemy#10312] [Usage Questions]
My query with SQLAlchemy is taking a long time to run a query with distinct. But when i execute a raw sql query using the sqlalchemy cursor, it is so much faster?

For exemple:
SQLAlchemy: session.query(Table.field).filter(Table.id == id).distinct()
Raw Sql: session.execute(text("SELECT DISTINCT mytable.field FROM myable WHERE mytable.id = :mytableid"), {"mytableid": id})

For my case, the results are:
Time Spent (150.34078192710876) seconds (SQLALCHEMY)
Time Spent (0.2926332950592041) seconds (RAW SQL)

**Comment by zzzeek** [comment]
your raw SQL is still going through SQLAlchemy, however you need to make sure you also fully fetch all rows and make sure result-fetching overhead is taken into account.  if this returns a million rows and you don't fetch in the second case, depending on driver you might see this large difference.

also, turn on echo=True and see exactly what SQL is being emitted for your ORM query, if some kind of joined inheritance rules or something are causing the query to be dramatically different.

Overall, follow the steps at https://docs.sqlalchemy.org/en/20/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application which will give you an exact answer to where the time is spent (but make sure you fetch all rows in both cases).

**Reply by gabrielmrts** [reply]
thanks zek, i gonna try these approachs
